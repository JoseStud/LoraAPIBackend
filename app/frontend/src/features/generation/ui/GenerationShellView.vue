<template>
  <div class="generation-shell grid gap-6 xl:grid-cols-[3fr_2fr]">
    <div class="generation-page-container flex flex-col gap-6">
      <slot
        name="header"
        :is-connected="isConnected"
        :show-history="showHistory"
        :toggle-history="emitToggleHistory"
        :clear-queue="emitClearQueue"
        :has-active-jobs="hasActiveJobs"
      >
        <div class="page-header">
          <div class="flex justify-between items-center">
            <div>
              <h1 class="page-title">Generation Studio</h1>
              <p class="page-subtitle">Generate images with AI-powered LoRA integration</p>
              <div class="mt-2 flex items-center text-sm text-gray-500 gap-2">
                <span
                  class="inline-flex h-2 w-2 rounded-full"
                  :class="isConnected ? 'bg-green-500' : 'bg-red-500'"
                ></span>
                <span v-if="isConnected">Live updates connected</span>
                <span v-else>Reconnecting to updates…</span>
              </div>
            </div>
            <div class="header-actions flex gap-2">
              <button
                data-testid="toggle-history"
                @click="emitToggleHistory()"
                class="btn btn-secondary btn-sm"
                :class="{ 'btn-primary': showHistory }"
                type="button"
              >
                <svg class="w-4 h-4 mr-2" fill="none" stroke="currentColor" viewBox="0 0 24 24">
                  <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M12 8v4l3 3m6-3a9 9 0 11-18 0 9 9 0 0 1180z"></path>
                </svg>
                History
              </button>
              <button
                data-testid="clear-queue"
                @click="emitClearQueue()"
                class="btn btn-secondary btn-sm"
                :disabled="!hasActiveJobs"
                type="button"
              >
                <svg class="w-4 h-4 mr-2" fill="none" stroke="currentColor" viewBox="0 0 24 24">
                  <path
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16"
                  ></path>
                </svg>
                Clear Queue
              </button>
            </div>
          </div>
        </div>
      </slot>

      <slot
        name="primary"
        :generation-parameter-form="GenerationParameterForm"
        :generation-active-jobs-list="GenerationActiveJobsList"
        :generation-results-gallery="GenerationResultsGallery"
        :generation-system-status-card="GenerationSystemStatusCard"
        :params="params"
        :is-generating="isGenerating"
        :active-jobs="activeJobs"
        :sorted-active-jobs="sortedActiveJobs"
        :recent-results="recentResults"
        :show-history="showHistory"
        :format-time="formatTime"
        :get-job-status-classes="getJobStatusClasses"
        :get-job-status-text="getJobStatusText"
        :can-cancel-job="canCancelJob"
        :system-status="systemStatus"
        :get-system-status-classes="getSystemStatusClasses"
        :update-params="emitUpdateParams"
        :start-generation="emitStartGeneration"
        :load-from-composer="emitLoadFromComposer"
        :use-random-prompt="emitUseRandomPrompt"
        :save-preset="emitSavePreset"
        :cancel-job="emitCancelJob"
        :refresh-results="emitRefreshResults"
        :reuse-parameters="emitReuseParameters"
        :delete-result="emitDeleteResult"
        :show-image-modal="emitShowImageModal"
      >
        <div class="grid grid-cols-1 lg:grid-cols-3 gap-6">
          <div class="lg:col-span-1">
            <GenerationParameterForm
              :params="params"
              :is-generating="isGenerating"
              @update:params="emitUpdateParams"
              @start-generation="emitStartGeneration"
              @load-from-composer="emitLoadFromComposer"
              @use-random-prompt="emitUseRandomPrompt"
              @save-preset="emitSavePreset"
            />
          </div>

          <div class="lg:col-span-1">
            <div class="space-y-6">
              <GenerationActiveJobsList
                :active-jobs="activeJobs"
                :sorted-active-jobs="sortedActiveJobs"
                :format-time="formatTime"
                :get-job-status-classes="getJobStatusClasses"
                :get-job-status-text="getJobStatusText"
                :can-cancel-job="canCancelJob"
                @cancel-job="emitCancelJob"
              />

              <GenerationSystemStatusCard
                :system-status="systemStatus"
                :get-system-status-classes="getSystemStatusClasses"
              />
            </div>
          </div>

          <div class="lg:col-span-1">
            <GenerationResultsGallery
              :recent-results="recentResults"
              :show-history="showHistory"
              :format-time="formatTime"
              @refresh-results="emitRefreshResults"
              @reuse-parameters="emitReuseParameters"
              @delete-result="emitDeleteResult"
              @show-image-modal="emitShowImageModal"
            />
          </div>
        </div>
      </slot>
    </div>

    <slot
      name="secondary"
      :system-status-card="SystemStatusCard"
      :job-queue="JobQueue"
      :system-status="systemStatus"
      :active-jobs="activeJobs"
      :clear-queue="emitClearQueue"
      :has-active-jobs="hasActiveJobs"
      :is-connected="isConnected"
    >
      <div class="flex flex-col gap-6">
        <SystemStatusCard variant="detailed" />
        <JobQueue :show-clear-completed="true" />
      </div>
    </slot>
  </div>

  <div
    v-if="showModal && selectedResult"
    class="fixed inset-0 z-50 overflow-y-auto"
    data-testid="result-modal"
    @click.self="emitHideImageModal()"
  >
    <div class="flex items-center justify-center min-h-screen pt-4 px-4 pb-20 text-center sm:block sm:p-0">
      <div class="fixed inset-0 bg-gray-500 bg-opacity-75" @click="emitHideImageModal"></div>

      <div class="inline-block align-bottom bg-white rounded-lg text-left overflow-hidden shadow-xl transform transition-all sm:my-8 sm:align-middle sm:max-w-lg sm:w-full">
        <img
          v-if="selectedResult?.image_url"
          :src="selectedResult.image_url"
          :alt="selectedResult?.prompt ?? 'Generated image'"
          class="w-full"
        >
        <div class="p-4">
          <h3 class="text-lg font-medium text-gray-900 mb-2">Generation Details</h3>
          <div class="space-y-1 text-sm text-gray-600">
            <div><strong>Prompt:</strong> {{ selectedResult?.prompt ?? 'Unknown prompt' }}</div>
            <div>
              <strong>Size:</strong>
              {{ selectedResult?.width ?? '—' }}×{{ selectedResult?.height ?? '—' }}
            </div>
            <div><strong>Steps:</strong> {{ selectedResult?.steps ?? '—' }}</div>
            <div><strong>CFG Scale:</strong> {{ selectedResult?.cfg_scale ?? '—' }}</div>
            <div><strong>Seed:</strong> {{ selectedResult?.seed ?? '—' }}</div>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { computed } from 'vue'

import GenerationActiveJobsList from '../components/GenerationActiveJobsList.vue'
import GenerationParameterForm from '../components/GenerationParameterForm.vue'
import GenerationResultsGallery from '../components/GenerationResultsGallery.vue'
import GenerationSystemStatusCard from '../components/GenerationSystemStatusCard.vue'
import JobQueue from '../components/JobQueue.vue'
import SystemStatusCard from '../components/system/SystemStatusCard.vue'
import type { GenerationFormState, SystemStatusState } from '@/types'
import type {
  QueueItemView,
  ReadonlyQueue,
  ReadonlyResults,
  ResultItemView,
} from '@/features/generation/orchestrator'
import type { DeepReadonly } from '@/utils/freezeDeep'

const props = defineProps<{
  isConnected: boolean
  showHistory: boolean
  params: GenerationFormState
  isGenerating: boolean
  activeJobs: ReadonlyQueue
  sortedActiveJobs: ReadonlyQueue
  recentResults: ReadonlyResults
  formatTime: (value?: string) => string
  getJobStatusClasses: (status: QueueItemView['status']) => string
  getJobStatusText: (status: QueueItemView['status']) => string
  canCancelJob: (job: QueueItemView) => boolean
  systemStatus: DeepReadonly<SystemStatusState>
  getSystemStatusClasses: (status?: string) => string
  showModal: boolean
  selectedResult: ResultItemView | null
}>()

const emit = defineEmits<{
  (event: 'toggle-history'): void
  (event: 'clear-queue'): void
  (event: 'update-params', value: GenerationFormState): void
  (event: 'start-generation'): void
  (event: 'load-from-composer'): void
  (event: 'use-random-prompt'): void
  (event: 'save-preset'): void
  (event: 'cancel-job', jobId: string): void
  (event: 'refresh-results'): void
  (event: 'reuse-parameters', result: ResultItemView): void
  (event: 'delete-result', resultId: string | number): void
  (event: 'show-image-modal', result: ResultItemView): void
  (event: 'hide-image-modal'): void
}>()

const params = computed(() => props.params)
const isGenerating = computed(() => props.isGenerating)
const activeJobs = computed(() => props.activeJobs)
const sortedActiveJobs = computed(() => props.sortedActiveJobs)
const recentResults = computed(() => props.recentResults)
const showHistory = computed(() => props.showHistory)
const systemStatus = computed(() => props.systemStatus)
const showModal = computed(() => props.showModal)
const selectedResult = computed(() => props.selectedResult)

const hasActiveJobs = computed(() => activeJobs.value.length > 0)

const emitToggleHistory = (): void => {
  emit('toggle-history')
}

const emitClearQueue = (): void => {
  emit('clear-queue')
}

const emitUpdateParams = (value: GenerationFormState): void => {
  emit('update-params', value)
}

const emitStartGeneration = (): void => {
  emit('start-generation')
}

const emitLoadFromComposer = (): void => {
  emit('load-from-composer')
}

const emitUseRandomPrompt = (): void => {
  emit('use-random-prompt')
}

const emitSavePreset = (): void => {
  emit('save-preset')
}

const emitCancelJob = (jobId: string): void => {
  emit('cancel-job', jobId)
}

const emitRefreshResults = (): void => {
  emit('refresh-results')
}

const emitReuseParameters = (result: ResultItemView): void => {
  emit('reuse-parameters', result)
}

const emitDeleteResult = (resultId: string | number): void => {
  emit('delete-result', resultId)
}

const emitShowImageModal = (result: ResultItemView): void => {
  emit('show-image-modal', result)
}

const emitHideImageModal = (): void => {
  emit('hide-image-modal')
}

const formatTime = props.formatTime
const getJobStatusClasses = props.getJobStatusClasses
const getJobStatusText = props.getJobStatusText
const canCancelJob = props.canCancelJob
const getSystemStatusClasses = props.getSystemStatusClasses
const isConnected = computed(() => props.isConnected)
</script>
