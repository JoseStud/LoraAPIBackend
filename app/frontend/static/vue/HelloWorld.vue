<template>
  <div class="p-4 bg-blue-100 border border-blue-300 rounded-lg">
    <h2 class="text-xl font-bold text-blue-800">Hello from Vue!</h2>
    <p class="text-blue-600">{{ message }}</p>
  </div>
  
</template>

<script setup>
import { ref } from 'vue';
const message = ref('This is our first Vue island.');
</script>

