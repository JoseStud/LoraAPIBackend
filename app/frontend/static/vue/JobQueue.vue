<template>
  <div v-if="isReady" class="card" :class="cardClass">
    <div class="card-header">
      <div class="flex items-center justify-between">
        <h3 class="card-title">{{ title }}</h3>
        <div class="flex items-center space-x-2">
          <span v-if="showJobCount" class="text-sm text-gray-500">{{ jobCountLabel }}</span>
          <button 
            v-if="showClearCompleted && hasJobs"
            @click="handleClearCompleted"
            class="btn btn-outline btn-sm"
            type="button"
          >
            Clear Completed
          </button>
        </div>
      </div>
    </div>
    <div class="card-body">
      <!-- Active Jobs -->
      <div class="space-y-3 max-h-96 overflow-y-auto">
        <template v-for="job in jobs" :key="job.id">
          <div class="border border-gray-200 rounded-lg p-3">
            <div class="flex items-start justify-between mb-2">
              <div class="flex-1 min-w-0">
                <div class="text-sm font-medium text-gray-900 truncate">
                  {{ job.name || job.prompt }}
                </div>
                <div class="text-xs text-gray-500">
                  {{ getJobDetailsText(job) }}
                </div>
              </div>
              <button 
                v-if="canCancelJob(job)"
                @click="handleCancelJob(job.id)" 
                class="text-gray-400 hover:text-red-500 ml-2"
                type="button"
                :disabled="isCancelling"
              >
                <svg class="w-4 h-4" fill="none" stroke="currentColor" viewBox="0 0 24 24">
                  <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M6 18L18 6M6 6l12 12"></path>
                </svg>
              </button>
            </div>
            
            <!-- Progress Bar -->
            <div class="space-y-1">
              <div class="flex justify-between text-xs">
                <span :class="getStatusColorClass(job.status)">{{ job.status }}</span>
                <span class="text-gray-600">{{ job.progress || 0 }}%</span>
              </div>
              <div class="progress-bar-bg">
                <div 
                  class="progress-bar-fg bg-blue-500" 
                  :style="{ width: `${job.progress || 0}%` }"
                ></div>
              </div>
              <div class="flex justify-between text-xs text-gray-500">
                <span>{{ formatDuration(job.startTime) }}</span>
                <span>{{ job.message || '' }}</span>
              </div>
            </div>
          </div>
        </template>
        
        <!-- Empty State -->
        <div v-if="!hasJobs" class="empty-state-container">
          <svg class="empty-state-icon" fill="none" stroke="currentColor" viewBox="0 0 24 24">
            <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M12 6V4m0 2a2 2 0 100 4m0-4a2 2 0 110 4m-6 8a2 2 0 100-4m0 4a2 2 0 100 4m0-4v2m0-6V4m6 6v10m6-2a2 2 0 100-4m0 4a2 2 0 100 4m0-4v2m0-6V4"></path>
          </svg>
          <div class="empty-state-title">{{ emptyStateTitle }}</div>
          <div class="empty-state-message">{{ emptyStateMessage }}</div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { ref, computed, onMounted, onUnmounted } from 'vue';
import { useApi } from './composables/useApi.js';

export default {
  name: 'JobQueue',
  props: {
    title: {
      type: String,
      default: 'Generation Queue'
    },
    emptyStateTitle: {
      type: String,
      default: 'No active generations'
    },
    emptyStateMessage: {
      type: String,
      default: 'Start a generation to see progress here'
    },
    cardClass: {
      type: String,
      default: ''
    },
    pollingInterval: {
      type: Number,
      default: 2000
    },
    disabled: {
      type: Boolean,
      default: false
    },
    showJobCount: {
      type: Boolean,
      default: true
    },
    showClearCompleted: {
      type: Boolean,
      default: false
    }
  },
  setup(props) {
    // Reactive state
    const isReady = ref(false);
    const isPolling = ref(false);
    const apiAvailable = ref(true);
    const isCancelling = ref(false);
    const pollingTimer = ref(null);

    // Access to global Alpine store - this maintains compatibility
    const store = window.Alpine?.store('app');
    
    // Job data - computed from the Alpine store for compatibility
    const jobs = computed(() => store?.activeJobs || []);
    const hasJobs = computed(() => jobs.value.length > 0);
    const jobCountLabel = computed(() => `${jobs.value.length} active`);

    // Utility functions
    const formatDuration = (startTime) => {
      const now = new Date();
      const start = new Date(startTime);
      const diffMs = now - start;
      const diffSecs = Math.floor(diffMs / 1000);
      
      if (diffSecs < 60) return `${diffSecs}s`;
      const diffMins = Math.floor(diffSecs / 60);
      if (diffMins < 60) return `${diffMins}m ${diffSecs % 60}s`;
      const diffHours = Math.floor(diffMins / 60);
      return `${diffHours}h ${diffMins % 60}m`;
    };

    const getStatusColorClass = (status) => {
      switch (status) {
        case 'running': return 'text-blue-600';
        case 'completed': return 'text-green-600';
        case 'failed': return 'text-red-600';
        case 'cancelled': return 'text-gray-600';
        default: return 'text-yellow-600';
      }
    };

    const getJobDetailsText = (job) => {
      if (job.params) {
        return `${job.params.width}x${job.params.height} â€¢ ${job.params.steps} steps`;
      }
      return 'Processing...';
    };

    const canCancelJob = (job) => {
      return job.status === 'running' || job.status === 'starting';
    };

    // Job management functions
    const handleClearCompleted = () => {
      if (!store) return;

      const completedJobs = store.activeJobs.filter(job => 
        job.status === 'completed' || job.status === 'failed'
      );
      
      completedJobs.forEach(job => {
        store.removeJob(job.id);
      });
    };

    const updateJobStatuses = async () => {
      if (!apiAvailable.value || !store) return;

      try {
        const backendUrl = window?.BACKEND_URL || '';
        const response = await fetch(`${backendUrl}/jobs/status`);
        
        if (response.ok) {
          const apiJobs = await response.json();
          
          // Update each job in the store
          apiJobs.forEach(apiJob => {
            const storeJob = store.activeJobs.find(j => j.jobId === apiJob.id);
            if (storeJob) {
              store.updateJob(storeJob.id, {
                status: apiJob.status,
                progress: apiJob.progress || 0,
                message: apiJob.message
              });
              
              // If job is complete, move to results
              if (apiJob.status === 'completed' && apiJob.result) {
                store.addResult(apiJob.result);
                store.removeJob(storeJob.id);
                store.addNotification('Generation completed!', 'success');
              } else if (apiJob.status === 'failed') {
                store.removeJob(storeJob.id);
                store.addNotification(`Generation failed: ${apiJob.error}`, 'error');
              }
            }
          });
        } else if (response.status === 404) {
          // API endpoint not implemented yet, stop polling
          apiAvailable.value = false;
          stopPolling();
        }
      } catch (error) {
        // Silently handle polling errors (like the Alpine version)
        console.debug('[JobQueue] Polling error:', error);
      }
    };

    const handleCancelJob = async (jobId) => {
      if (!store || isCancelling.value) return;

      try {
        isCancelling.value = true;
        const job = store.activeJobs.find(j => j.id === jobId);
        
        if (job && job.jobId) {
          const backendUrl = window?.BACKEND_URL || '';
          const response = await fetch(`${backendUrl}/jobs/${job.jobId}/cancel`, {
            method: 'POST'
          });
          
          if (response.ok) {
            store.removeJob(jobId);
            store.addNotification('Job cancelled', 'info');
          } else {
            store.addNotification('Failed to cancel job', 'error');
          }
        }
      } catch (error) {
        store.addNotification('Failed to cancel job', 'error');
      } finally {
        isCancelling.value = false;
      }
    };

    // Polling management
    const startPolling = () => {
      if (!apiAvailable.value || pollingTimer.value || props.disabled) {
        return;
      }

      console.log('[JobQueue] Starting job polling every', props.pollingInterval, 'ms');
      
      pollingTimer.value = setInterval(async () => {
        if (isPolling.value) {
          console.debug('[JobQueue] Polling in progress, skipping new request');
          return;
        }
        
        if (!apiAvailable.value) {
          stopPolling();
          return;
        }
        
        try {
          isPolling.value = true;
          await updateJobStatuses();
        } catch (error) {
          console.error('[JobQueue] Job status update failed:', error);
        } finally {
          isPolling.value = false;
        }
      }, props.pollingInterval);
    };

    const stopPolling = () => {
      if (pollingTimer.value) {
        clearInterval(pollingTimer.value);
        pollingTimer.value = null;
        console.log('[JobQueue] Polling stopped');
      }
      isPolling.value = false;
    };

    // Lifecycle
    onMounted(() => {
      console.log('[JobQueue] Vue component mounted');
      isReady.value = true;
      
      if (!props.disabled) {
        startPolling();
      }
    });

    onUnmounted(() => {
      console.log('[JobQueue] Vue component unmounting, cleaning up');
      stopPolling();
    });

    return {
      isReady,
      jobs,
      hasJobs,
      jobCountLabel,
      isCancelling,
      formatDuration,
      getStatusColorClass,
      getJobDetailsText,
      canCancelJob,
      handleCancelJob,
      handleClearCompleted
    };
  }
};
</script>