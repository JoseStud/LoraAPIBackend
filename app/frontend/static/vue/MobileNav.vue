<template>
  <div class="mobile-nav-root">
    <!-- Mobile Navigation Toggle -->
    <button
      class="mobile-nav-toggle lg:hidden touch-target"
      :class="{ 'mobile-safe-top': true }"
      :aria-expanded="isOpen"
      aria-controls="mobile-navigation"
      aria-label="Toggle mobile navigation menu"
      @click="toggleMenu"
    >
      <svg class="w-6 h-6" fill="none" stroke="currentColor" viewBox="0 0 24 24">
        <path
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          :d="isOpen ? 'M6 18L18 6M6 6l12 12' : 'M4 6h16M4 12h16M4 18h16'"
        />
      </svg>
      <span class="sr-only">{{ isOpen ? 'Close menu' : 'Open menu' }}</span>
    </button>

    <!-- Mobile Navigation Overlay -->
    <div
      v-show="isOpen"
      @click="closeMenu"
      class="mobile-nav-overlay"
    />

    <!-- Mobile Navigation Menu -->
    <nav
      class="mobile-nav-menu"
      id="mobile-navigation"
      :class="isOpen ? 'open' : 'closed'"
      role="navigation"
      aria-label="Mobile navigation"
      :aria-hidden="!isOpen"
    >
      <div class="mobile-nav-header mobile-safe-top">
        <h2 class="text-lg font-semibold text-white">LoRA Manager</h2>
        <button
          @click="closeMenu"
          class="mobile-nav-close touch-target"
          aria-label="Close navigation menu"
        >
          <svg class="w-6 h-6" fill="none" stroke="currentColor" viewBox="0 0 24 24">
            <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M6 18L18 6M6 6l12 12" />
          </svg>
        </button>
      </div>

      <div class="mobile-nav-links">
        <a
          v-for="item in items"
          :key="item.path"
          :href="item.path"
          @click="closeMenu"
          class="mobile-nav-link"
          :class="{ 'active': isActive(item.path) }"
        >
          <span class="inline-flex items-center">
            <svg class="w-5 h-5 mr-3 inline" fill="none" stroke="currentColor" viewBox="0 0 24 24">
              <path v-if="item.icon === 'dashboard'" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M3 12l2-2m0 0l7-7 7 7M13 5v6a2 2 0 002 2h6" />
              <path v-else-if="item.icon === 'grid'" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M4 6h7v7H4V6zm9 0h7v7h-7V6zM4 15h7v7H4v-7zm9 0h7v7h-7v-7z" />
              <path v-else-if="item.icon === 'spark'" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M5 3l7 7-7 7M13 5l6 6-6 6" />
              <path v-else-if="item.icon === 'compose'" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M11 5H6a2 2 0 00-2 2v11a2 2 0 002 2h11a2 2 0 002-2v-5M15 3l6 6M8 13h4" />
              <path v-else-if="item.icon === 'wand'" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M15 7l-1 4-4 1 1-4 4-1zM3 21l6-6" />
              <path v-else-if="item.icon === 'admin'" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M12 12c2.21 0 4-1.79 4-4S14.21 4 12 4 8 5.79 8 8s1.79 4 4 4z M4 20a8 8 0 0116 0" />
              <path v-else-if="item.icon === 'bars'" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M9 19v-6a2 2 0 00-2-2H5a2 2 0 00-2 2v6m8 0V9a2 2 0 012-2h2a2 2 0 012 2v10" />
              <path v-else stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M4 6h16M4 12h16M4 18h16" />
            </svg>
            {{ item.label }}
          </span>
        </a>
      </div>

      <div class="p-4 border-t border-gray-200 mobile-safe-bottom">
        <div class="text-xs text-gray-500">LoRA Manager v2.1</div>
      </div>
    </nav>
  </div>
  
</template>

<script setup>
import { ref, onMounted, onBeforeUnmount } from 'vue';

const isOpen = ref(false);

const items = [
  { path: '/', label: 'Dashboard', icon: 'dashboard' },
  { path: '/loras', label: 'LoRAs', icon: 'grid' },
  { path: '/recommendations', label: 'Recommendations', icon: 'spark' },
  { path: '/compose', label: 'Compose', icon: 'compose' },
  { path: '/generate', label: 'Generate', icon: 'wand' },
  { path: '/admin', label: 'Admin', icon: 'admin' },
  { path: '/analytics', label: 'Analytics', icon: 'bars' },
  { path: '/import-export', label: 'Import/Export', icon: 'grid' },
];

const toggleMenu = () => { isOpen.value = !isOpen.value; };
const closeMenu = () => { isOpen.value = false; };

const isActive = (path) => {
  try {
    return (window?.location?.pathname || '').replace(/\/+$/, '') === path;
  } catch { return false; }
};

const onKeydown = (e) => {
  if (e.key === 'Escape') closeMenu();
};

onMounted(() => {
  document.addEventListener('keydown', onKeydown);
});

onBeforeUnmount(() => {
  document.removeEventListener('keydown', onKeydown);
});
</script>

<style scoped>
/* Rely on existing app styles; no overrides here */
</style>
