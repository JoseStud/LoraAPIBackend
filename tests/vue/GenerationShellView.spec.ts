import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import { defineComponent } from 'vue'

import GenerationShellView from '@/features/generation/ui/GenerationShellView.vue'
import type { GenerationFormState, GenerationJob, GenerationResult, SystemStatusState } from '@/types'
import type { ReadonlyResults, ResultItemView } from '@/features/generation/orchestrator'

const createComponentStub = (name: string, props: string[], emits: string[] = []) =>
  defineComponent({
    name,
    props,
    emits,
    template: `<div class="${name}-stub"><slot /></div>`,
  })

const GenerationParameterFormStub = createComponentStub(
  'GenerationParameterForm',
  ['params', 'isGenerating'],
  ['update:params', 'start-generation', 'load-from-composer', 'use-random-prompt', 'save-preset'],
)

const GenerationActiveJobsListStub = createComponentStub(
  'GenerationActiveJobsList',
  ['activeJobs', 'sortedActiveJobs', 'formatTime', 'getJobStatusClasses', 'getJobStatusText', 'canCancelJob'],
  ['cancel-job'],
)

const GenerationResultsGalleryStub = createComponentStub(
  'GenerationResultsGallery',
  ['recentResults', 'showHistory', 'formatTime'],
  ['refresh-results', 'reuse-parameters', 'delete-result', 'show-image-modal'],
)

const GenerationSystemStatusCardStub = createComponentStub(
  'GenerationSystemStatusCard',
  ['systemStatus', 'getSystemStatusClasses'],
)

const JobQueueStub = createComponentStub('JobQueue', ['showClearCompleted'])

const SystemStatusCardStub = createComponentStub('SystemStatusCard', ['variant'])

const params: GenerationFormState = {
  prompt: 'Example prompt',
  negative_prompt: '',
  steps: 20,
  sampler_name: 'Euler',
  cfg_scale: 7,
  width: 512,
  height: 512,
  seed: -1,
  batch_size: 1,
  batch_count: 1,
}

const createJob = (overrides: Partial<GenerationJob> = {}): GenerationJob => ({
  uiId: 'job-1',
  backendId: 'backend-1',
  status: 'queued',
  prompt: 'Job prompt',
  id: 'job-1',
  created_at: new Date().toISOString(),
  ...overrides,
}) as GenerationJob

const result: GenerationResult = {
  id: 'result-1',
  prompt: 'Result prompt',
  width: 512,
  height: 512,
  steps: 20,
  cfg_scale: 7,
  seed: 123,
  image_url: '/image.png',
}

const recentResults: ReadonlyResults = Object.freeze([{ ...result } as ResultItemView]) as ReadonlyResults

const systemStatus: SystemStatusState = {
  status: 'healthy',
  queue_length: 1,
  gpu_available: true,
  active_workers: 1,
  gpu_status: 'Ready',
  memory_total: 16,
  memory_used: 8,
  disk_total: 512,
  disk_used: 128,
  uptime_seconds: 100,
  last_job_started_at: null,
  last_job_completed_at: null,
  api_available: true,
}

const baseProps = () => ({
  isConnected: true,
  showHistory: false,
  params,
  isGenerating: false,
  activeJobs: [createJob()],
  sortedActiveJobs: [createJob()],
  recentResults,
  formatTime: (value?: string) => value ?? 'Unknown',
  getJobStatusClasses: () => 'status-class',
  getJobStatusText: () => 'Status',
  canCancelJob: () => true,
  systemStatus,
  getSystemStatusClasses: () => 'system-status-class',
  showModal: false,
  selectedResult: null as GenerationResult | null,
})

describe('GenerationShellView', () => {
  const global = {
    stubs: {
      GenerationParameterForm: GenerationParameterFormStub,
      GenerationActiveJobsList: GenerationActiveJobsListStub,
      GenerationResultsGallery: GenerationResultsGalleryStub,
      GenerationSystemStatusCard: GenerationSystemStatusCardStub,
      JobQueue: JobQueueStub,
      SystemStatusCard: SystemStatusCardStub,
    },
  }

  it('renders default layout', () => {
    const wrapper = mount(GenerationShellView, {
      props: baseProps(),
      global,
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('emits toggle-history when history button clicked', async () => {
    const wrapper = mount(GenerationShellView, {
      props: baseProps(),
      global,
    })

    await wrapper.get('[data-testid="toggle-history"]').trigger('click')

    expect(wrapper.emitted('toggle-history')).toHaveLength(1)
  })

  it('emits clear-queue when button clicked', async () => {
    const wrapper = mount(GenerationShellView, {
      props: baseProps(),
      global,
    })

    await wrapper.get('[data-testid="clear-queue"]').trigger('click')

    expect(wrapper.emitted('clear-queue')).toHaveLength(1)
  })

  it('renders modal when showModal is true and emits hide-image-modal on overlay click', async () => {
    const wrapper = mount(GenerationShellView, {
      props: {
        ...baseProps(),
        showModal: true,
        selectedResult: result,
      },
      global,
    })

    expect(wrapper.find('[data-testid="result-modal"]').exists()).toBe(true)

    await wrapper.get('[data-testid="result-modal"]').trigger('click')

    expect(wrapper.emitted('hide-image-modal')).toHaveLength(1)
  })
})
