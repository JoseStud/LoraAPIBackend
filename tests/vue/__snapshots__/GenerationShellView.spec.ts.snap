// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GenerationShellView > renders default layout 1`] = `
"<div class="generation-shell grid gap-6 xl:grid-cols-[3fr_2fr]">
  <div class="generation-page-container flex flex-col gap-6">
    <div class="grid grid-cols-1 lg:grid-cols-3 gap-6">
      <div class="lg:col-span-1">
        <div class="GenerationParameterForm-stub"></div>
      </div>
      <div class="lg:col-span-1">
        <div class="space-y-6">
          <div class="GenerationActiveJobsList-stub"></div>
          <div class="GenerationSystemStatusCard-stub"></div>
        </div>
      </div>
      <div class="lg:col-span-1">
        <div class="GenerationResultsGallery-stub"></div>
      </div>
    </div>
  </div>
  <div class="generation-secondary flex flex-col gap-6">
    <div class="flex flex-col gap-6">
      <div class="SystemStatusCard-stub"></div>
      <div class="JobQueue-stub"></div>
    </div>
  </div>
</div>
<!--v-if-->"
`;
